src/charm.py

```diff
@@ -481,134 +43,61 @@ def __init__(self, *args) -> None:  # type: ignore[no-untyped-def]
             args: Variable list of positional arguments passed to the parent constructor.
         """
         super().__init__(*args)
-        self.framework.observe(self.on.config_changed, self._on_config_changed_with_warning)
+        if not self.unit.is_leader():
+            self.unit.status = WaitingStatus(
+                "follower unit is idling, "
+                "remove follower units using `juju scale-application {self.app.name} 1`"
+            )
+            return
+
+        self._ingress_options: Optional[IngressOption] = None
+
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
         self.framework.observe(self.on.describe_ingresses_action, self._describe_ingresses_action)
-        self.framework.observe(self.on.start, self._on_start)
+        self.framework.observe(self.on.start, self._on_config_changed)
 
-        # 'ingress' relation handling.
-        self.ingress = IngressProvides(self)
-        # When the 'ingress' is ready to configure, do so.
-        self.framework.observe(self.on.ingress_available, self._on_config_changed_with_warning)
-        self.framework.observe(self.on.ingress_broken, self._on_ingress_broken)
+        self._ingress_provider = IngressPerAppProvider(charm=self)
```

Comment:
perhaps move this above to have all the observe together

Reply:
Updated, thanks!